ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs

```diff
@@ -709,16 +709,40 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = electric $ do
         curTip      = castPoint $ AF.headPoint curChain
         curHead     = AF.headAnchor curChain
 
-    -- Either frag extends loe or loe extends frag
+    -- | Trim the given candidate fragment to respect the LoE.
     --
-    -- PRECONDITION: @AF.withinFragmentBounds (AF.anchorPoint frag) loe@
-    followsLoEFrag :: LoE (AnchoredFragment (Header blk))
-                   -> AnchoredFragment (Header blk)
-                   -> Bool
-    followsLoEFrag LoEDisabled _ = True
-    followsLoEFrag (LoEEnabled loe) frag =
-         AF.withinFragmentBounds (AF.headPoint loe) frag
-      || AF.withinFragmentBounds (AF.headPoint frag) loe
+    -- The returned fragment is such that:
+    --
+    -- - It is a prefix of the given fragment.
+    -- - If it contains the tip of the LoE fragment, then it contains at most
+    --   @k@ block after it.
+    -- - If it does not contain the tip of the LoE fragment, then it is included
+    --   in the LoE fragment.
+    --
+    -- The fragment is represented by the current chain and a diff with that
+    -- current chain. It is tempting to only consider the suffix of the diff,
+    -- but that would be incorrect, because the diff might not intersect with
+    -- the LoE fragment, because the diff suffix is anchored somewhere on the
+    -- current chain and LoE frag's tip might be older than that anchor.
+    --
+    -- PRECONDITIONS:
+    --
+    -- 1. The given 'ChainDiff' can apply on top of the given 'ChainAndLedger'.
+    -- 2. The LoE fragment intersects with the current selection.
```

Comment:
Assert these with `assertWithMsg`?

Reply:
Enforced preconditions in faece5de1f796ea9dd94664eca11979003c0446f.